### PR TITLE
TRUNK-6203: Global properties access should be privileged

### PR DESCRIPTION
**Pull request by Seremba**
## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java [hunk 1]
```diff
@@ -144,34 +147,55 @@ public SortedMap<String, String> getSystemVariables() throws APIException {
...
 	@Override
 	@Transactional(readOnly = true)
 	public String getGlobalProperty(String propertyName) throws APIException {
-		// This method should not have any authorization check
 		if (propertyName == null) {
 			return null;
 		}
+		try {
+			Context.addProxyPrivilege(PrivilegeConstants.GET_GLOBAL_PROPERTIES);
+			if (!Context.isAuthenticated()) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java:159]: The authentication check is automatically done by the framework. So not need for all this logic here. Not so?

Comment (Seremba) [reply api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java:159]: @dkayiwa you must be right. Let me remove it

**Comment by dkayiwa** [comment]
Whenever you are ready for review, you need to change the pull request from draft.

**Comment by Seremba** [comment]
> Whenever you are ready for review, you need to change the pull request from draft.

Thank you, I have done as suggested.

**Comment by dkayiwa** [comment]
Are you able to successfully run openmrs-core with these changes?
If yes, can you also load the legacyui module and confirm that you can log in successfully?

**Comment by Seremba** [comment]
Yes, I am able to run openmrs-core successfully and I have been able to login into the system. With loading the legacy UI module, do you mean deleting it and uploading its omod file again?

**Comment by dkayiwa** [comment]
Can you run core, with these changes, together with all the reference application 2.x modules?

**Comment by Seremba** [comment]
Yes! However, there is something I have just discovered. When I run the core with `mvn openmrs-sdk:run -DserverId=server` things work perfectly well and I am able login into the system. But after like 10 to 15 minutes, without doing anything on the system, I get this error https://pastebin.com/vdCkt50T. When I interact with the system, like say I add a new patient, the error then disappears and everything works normally. The error does not affect any operations thus far but it is there. uhm!

**Comment by dkayiwa** [comment]
Clone the addresshierarchy module and add the requested proxy privilege as in the error log. Then compile and update your reference application. Do the errors disappear?

**Comment by Seremba** [comment]
No. they don't appear anymore! So, how do I have to create a PR for the changes in addresshierarchy module too?

File: api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java [hunk 2]
```diff
@@ -144,34 +147,62 @@ public SortedMap<String, String> getSystemVariables() throws APIException {
...
+				throw new APIException("GlobalProperty.error.privilege.required.view",
+				        new Object[] { gp.getViewPrivilege().getPrivilege(), propertyName });
 			}
 		} else {
 			return null;
 		}
+		
+	}
+	
+	String[] anonymousArray = { "owa.appBaseUrl", "login.url", "spa.baseUrl", "timezone.conversions", "default_theme",
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java:186]: All of these should be moved to the top of the file with the remainder of the properties.

File: api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java [hunk 3]
```diff
@@ -144,34 +147,62 @@ public SortedMap<String, String> getSystemVariables() throws APIException {
...
+	        "gzip.enabled", "default_locale" };
+	
+	private final Set<String> anonymouslyAccessibleProperties = new HashSet<>(Arrays.asList(anonymousArray));
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java:189]: It seems kind of silly to create an array, cast it into a list and then a set, when we could just do:

```java
private final static Set<String> anonymouslyAccessibleProperties = new HashSet<>();
static {
  Collections.addAll(anonymouslyAccessibleProperties, "owa.appBaseUrl", "login.url", "spa.baseUrl", "timezone.conversions", "default_theme", "gzip.enabled", "default_locale")
}
```

**Review by ibacher — COMMENTED** [review]
Design-wise, I guess a hard-coded list is ok, but it would be even better if modules could actually declare which GP's should be anonymously accessible.

**Comment by Seremba** [comment]
> Design-wise, I guess a hard-coded list is ok, but it would be even better if modules could actually declare which GP's should be anonymously accessible.

@ibacher wouldn't this work best if the list was quite big?

**Comment by Seremba** [comment]
> Clone the addresshierarchy module and add the requested proxy privilege as in the error log. Then compile and update your reference application. Do the errors disappear?

Hi @dkayiwa, do I have to make a PR for the changes in addresshierarchy module too? It has been an hour and the error doesn't seem to appear anymore.

**Comment by dkayiwa** [comment]
> > Clone the addresshierarchy module and add the requested proxy privilege as in the error log. Then compile and update your reference application. Do the errors disappear?
> 
> Hi @dkayiwa, do I have to make a PR for the changes in addresshierarchy module too? It has been an hour and the error doesn't seem to appear anymore.

Yes create a jira ticket under that project and raise a pull request.

**Comment by Seremba** [comment]
Unfortunately, the error reappeared after an hour! Wow cc @dkayiwa

**Comment by dkayiwa** [comment]
> Unfortunately, the error reappeared after an hour! Wow cc @dkayiwa

Welcome to coding. 😊

**Comment by Seremba** [comment]
Hi @dkayiwa, addresshierarchy module uses `<openMRSVersion>1.10.6</openMRSVersion>` and it misses `Get Global Properties` in its PrivilegeConstants file. I can't add it. It is a read-only file. Is there a way you recommend going about this?

**Comment by dkayiwa** [comment]
Use the actual string value `Get Global Properties` instead of the constant.

**Comment by Seremba** [comment]
Thank you, I had tried that earlier, but the problem still was there! Maybe I had done something wrong. Let me try again

**Comment by Seremba** [comment]
> > > Clone the addresshierarchy module and add the requested proxy privilege as in the error log. Then compile and update your reference application. Do the errors disappear?
> > 
> > 
> > Hi @dkayiwa, do I have to make a PR for the changes in addresshierarchy module too? It has been an hour and the error doesn't seem to appear anymore.
> 
> Yes create a jira ticket under that project and raise a pull request.

@dkayiwa https://github.com/openmrs/openmrs-module-addresshierarchy/pull/41

**Comment by Seremba** [comment]
> > > Clone the addresshierarchy module and add the requested proxy privilege as in the error log. Then compile and update your reference application. Do the errors disappear?
> > 
> > 
> > Hi @dkayiwa, do I have to make a PR for the changes in addresshierarchy module too? It has been an hour and the error doesn't seem to appear anymore.
> 
> Yes create a jira ticket under that project and raise a pull request.

HI @dkayiwa, here is the PR under addresshierarchy project https://github.com/openmrs/openmrs-module-addresshierarchy/pull/41

**Comment by Seremba** [comment]
@dkayiwa , @ibacher is there anything more I am still meant to do with this PR?

**Comment by Seremba** [comment]
Hi @dkayiwa I am kindly requesting for the review of this PR
